File: go/driver/headgateredis/lua/enqueue.lua
-- Batch enqueue, atomic and all-or-nothing (a script IS the transaction on Redis).
-- Same contract as the Postgres enqueue: crash quarantine quarantined fingerprints rejected, job uniqueness
-- one duplicate semantic carrying the winner's id, store time for enqueued_at and for
-- "scheduled_at_ms = 0 means now".
--
-- Uniqueness: LIFECYCLE keys live at {p}:unique:{key} -> job id, released when the job
-- reaches a terminal state. THROTTLE keys live at {p}:uniquet:{key} with PX = window —
-- Redis TTL is the store's own clock, so "released by the clock" needs no sweeper.
--
-- KEYS[1] prefix
-- ARGV[1] n, then 17 args per job:
--   id, kind, schema_version, payload, queue, partition_key, rate_class, fingerprint,
--   priority, max_attempts, scheduled_at_ms, timeout_ms, deadline_ms, retention_ms,
--   unique_key(''=none), unique_window_ms, unique_states
-- then, OPTIONALLY, a TRAILING BLOCK of n more args: headers[i] as a JSON object
--   ('' or absent = none). telemetry and trace context regression revision — the RESERVED `traceparent`/`tracestate`
--   keys live in here and mean nothing to this script.
-- then, OPTIONALLY, n weights. Missing/zero means the backward-compatible default 1.
-- then, OPTIONALLY, n periodic schedule ids and n periodic tick times. Missing means
-- ordinary enqueue. These are typed hash fields, never opaque-header conventions.
-- then, OPTIONALLY, n unique-replacement masks. Bits are payload=1, scheduled_at=2,
-- priority=4, max_attempts=8; then n debounce windows, n pending flags, n tag arrays,
-- and n sticky worker ids. Sticky routing is strict; '' means any worker.
--
-- The headers ride in a TRAILING BLOCK rather than as an 18th per-job field ON PURPOSE:
-- every existing index expression above is `2 + i * F + k`, so widening the stride F
-- would silently move all seventeen of them. A trailing block leaves the three passes
-- below byte-identical and makes the growth strictly additive — including for a caller
-- that does not send the block at all, where ARGV[...] is simply nil.
-- Returns {'OK', n} | {'DUP', existing_id} | {'DUPR', existing_id} | {'QUAR', fingerprint} | {'IDC', id}
--       | {'BACK', queue, limit, current, incoming} | {'ERR', msg}
local p = KEYS[1]
local t = redis.call('TIME')
local now = tonumber(t[1]) * 1000 + math.floor(tonumber(t[2]) / 1000)
local n = tonumber(ARGV[1])
local F = 17

local function pending_key(queue, part, sticky)
  local k = p..':pending:'..queue..':'..part
  if sticky and sticky ~= '' then return k..':worker:'..sticky end
  return k
end
local function route_token(sticky)
  if sticky and sticky ~= '' then return sticky end
  return '*'
end
local function add_waiting(id, queue, part, sticky, score)
  redis.call('ZADD', pending_key(queue, part, sticky), score, id)
  redis.call('SADD', p..':pending-routes:'..queue..':'..part, route_token(sticky))
  redis.call('SADD', p..':parts:'..queue, part)
end

-- idempotent enqueue identity ID pass (regression revision): the strict caller-supplied id contract, run over the WHOLE
-- batch before any other check so all three backends classify a mixed batch the same way.
-- An id whose hash exists with matching CONTENT (kind, content fingerprinting fingerprint, queue) is
-- skipped entirely — idempotent success, the thing that makes the API's Idempotency-Key
-- replay safe, and skipping means no re-write, no counter, no wakeup, and no unique-key
-- check that would otherwise find the job conflicting with ITSELF. Different content is
-- {'IDC', id} and rejects the whole batch. A terminal job's hash still exists, so id
-- reuse follows retention eviction. (The equivalent of the SQL backends' pre-check
-- SELECT; here the script IS the transaction, so no race window exists at all.)
local skip = {}
for i = 0, n - 1 do
  local b = 2 + i * F
  local id, kind, queue, fp = ARGV[b], ARGV[b + 1], ARGV[b + 4], ARGV[b + 7]
  if id == '' then return {'ERR', 'envelope id must not be empty'} end
  local jk = p..':job:'..id
  if redis.call('EXISTS', jk) == 1 then
    local ex = redis.call('HMGET', jk, 'kind', 'fingerprint', 'queue')
    if ex[1] == kind and ex[2] == fp and ex[3] == queue then
      skip[i] = true
    else
      return {'IDC', id}
    end
  end
end

-- Validate pass: nothing is written unless the whole batch is admissible.
local seen_keys = {}
local written = 0
local demand = {}
for i = 0, n - 1 do
  if not skip[i] then
    local b = 2 + i * F
    local id, queue, fp, uk = ARGV[b], ARGV[b + 4], ARGV[b + 7], ARGV[b + 14]
    written = written + 1
    demand[queue] = (demand[queue] or 0) + 1
    if fp ~= '' and redis.call('SISMEMBER', p..':quarantine', fp) == 1 then
      return {'QUAR', fp}
    end
    if uk ~= '' then
      if seen_keys[uk] then return {'DUP', seen_keys[uk]} end
      seen_keys[uk] = id
      local uw = tonumber(ARGV[b + 15])
      local holder
      if uw > 0 then holder = redis.call('GET', p..':uniquet:'..uk)
      else holder = redis.call('GET', p..':unique:'..uk) end
      if holder then
        local mask = tonumber(ARGV[2 + n * F + 4 * n + i] or '0')
        local debounce = tonumber(ARGV[2 + n * F + 5 * n + i] or '0')
        if mask > 0 or debounce > 0 then
          local jk = p..':job:'..holder
          local state = redis.call('HGET', jk, 'state')
          local mutable = state == 'pending' or state == 'scheduled' or state == 'available' or state == 'retryable'
          local changed = false
          if mutable then
            if debounce > 0 then
              local oldfp = redis.call('HGET', jk, 'fingerprint') or ''
              local route = redis.call('HMGET', jk, 'queue', 'partition_key', 'sticky_worker')
              local oldtags = cjson.decode(redis.call('HGET', jk, 'tags') or '[]')
              for _, tag in ipairs(oldtags) do redis.call('SREM', p..':tag:'..tag, holder) end
              local tagsjson = ARGV[2 + n * F + 7 * n + i] or '[]'
              local tags = cjson.decode(tagsjson)
              for _, tag in ipairs(tags) do redis.call('SADD', p..':tag:'..tag, holder) end
              redis.call('ZREM', p..':idx:'..route[1]..':'..state, holder)
              redis.call('ZREM', p..':avail:'..route[1]..':'..route[2], holder)
              redis.call('HSET', jk, 'schema_version', ARGV[b + 2], 'payload', ARGV[b + 3],
                         'fingerprint', ARGV[b + 7], 'tags', tagsjson,
                         'state', 'scheduled', 'scheduled_at_ms', now + debounce)
              if oldfp ~= '' then redis.call('SREM', p..':fpi:'..oldfp, holder) end
              if ARGV[b + 7] ~= '' then redis.call('SADD', p..':fpi:'..ARGV[b + 7], holder) end
              add_waiting(holder, route[1], route[2], route[3] or '', now + debounce)
              redis.call('ZADD', p..':sched', now + debounce, holder)
              redis.call('ZADD', p..':idx:'..route[1]..':scheduled', now + debounce, holder)
              changed = true
            end
            if mask % 2 >= 1 then
              local oldfp = redis.call('HGET', jk, 'fingerprint') or ''
              redis.call('HSET', jk, 'schema_version', ARGV[b + 2], 'payload', ARGV[b + 3],
                         'fingerprint', ARGV[b + 7])
              if oldfp ~= '' then redis.call('SREM', p..':fpi:'..oldfp, holder) end
              if ARGV[b + 7] ~= '' then redis.call('SADD', p..':fpi:'..ARGV[b + 7], holder) end
              changed = true
            end
            if math.floor(mask / 2) % 2 >= 1 and state == 'scheduled' then
              local sched = tonumber(ARGV[b + 10]); if sched == 0 then sched = now end
              redis.call('HSET', jk, 'scheduled_at_ms', sched)
              local route = redis.call('HMGET', jk, 'queue', 'partition_key', 'sticky_worker')
              add_waiting(holder, route[1], route[2], route[3] or '', sched)
              redis.call('ZADD', p..':sched', sched, holder)
              redis.call('ZADD', p..':idx:'..route[1]..':scheduled', sched, holder)
              changed = true
            end
            if math.floor(mask / 4) % 2 >= 1 then
              redis.call('HSET', jk, 'priority', ARGV[b + 8])
              changed = true
            end
            if math.floor(mask / 8) % 2 >= 1 then
              redis.call('HSET', jk, 'max_attempts', ARGV[b + 9])
              changed = true
            end
          end
          if changed then return {'DUPR', holder} end
        end
        return {'DUP', holder}
      end
    end
  end
end

-- Exact producer depth is two scalar counters, never a scan. The script is Redis's
-- transaction and therefore serializes the verdict with every increment below.
local demand_queues = {}
for queue, _ in pairs(demand) do demand_queues[#demand_queues + 1] = queue end
table.sort(demand_queues)
for _, queue in ipairs(demand_queues) do
  local h = redis.call('HMGET', p..':enqueue:'..queue, 'limit', 'entered', 'exited')
  if h[1] then
    local limit = tonumber(h[1])
    local current = math.max(0, tonumber(h[2] or '0') - tonumber(h[3] or '0'))
    local incoming = demand[queue]
    if current + incoming > limit then
      return {'BACK', queue, tostring(limit), tostring(current), tostring(incoming)}
    end
  end
end

-- Write pass.
local woken = {}
for i = 0, n - 1 do
 if not skip[i] then
  local b = 2 + i * F
  local id, kind, queue, part = ARGV[b], ARGV[b + 1], ARGV[b + 4], ARGV[b + 5]
  local sched = tonumber(ARGV[b + 10])
  local debounce = tonumber(ARGV[2 + n * F + 5 * n + i] or '0')
  if debounce > 0 then sched = now + debounce elseif sched == 0 then sched = now end
  local state = 'available'
  if tonumber(ARGV[2 + n * F + 6 * n + i] or '0') == 1 then state = 'pending'
  elseif sched > now then state = 'scheduled' end
  local tagsjson = ARGV[2 + n * F + 7 * n + i] or '[]'
  local sticky = ARGV[2 + n * F + 8 * n + i] or ''
  redis.call('HSET', p..':job:'..id,
    'kind', kind, 'schema_version', ARGV[b + 2], 'payload', ARGV[b + 3],
    'queue', queue, 'partition_key', part, 'rate_class', ARGV[b + 6],
    'weight', tostring(math.max(1, tonumber(ARGV[2 + n * F + n + i] or '1'))),
    'rate_charge', 0,
    'fingerprint', ARGV[b + 7], 'priority', ARGV[b + 8],
    'attempt', 0, 'crash_attempt', 0, 'max_attempts', ARGV[b + 9],
    'enqueued_at_ms', now, 'scheduled_at_ms', sched,
    'timeout_ms', ARGV[b + 11], 'deadline_ms', ARGV[b + 12],
    'retention_ms', ARGV[b + 13], 'unique_key', ARGV[b + 14],
    'unique_window_ms', ARGV[b + 15], 'unique_states', ARGV[b + 16],
    'periodic_schedule_id', ARGV[2 + n * F + 2 * n + i] or '',
    'periodic_tick_ms', ARGV[2 + n * F + 3 * n + i] or '0',
    'tags', tagsjson, 'sticky_worker', sticky,
    'state', state, 'fence', 0, 'errors', '[]')
  -- telemetry and trace context regression revision: opaque headers from the trailing block. Written only when present,
  -- so a header-less job's hash is byte-identical to what it was before this existed.
  local hdr = ARGV[2 + n * F + i]
  if hdr and hdr ~= '' and hdr ~= '{}' then
    redis.call('HSET', p..':job:'..id, 'headers', hdr)
  end
  if state ~= 'pending' then add_waiting(id, queue, part, sticky, sched) end
  redis.call('ZADD', p..':metricparts:'..queue, now, part)
  redis.call('SADD', p..':queues', queue)
  redis.call('HINCRBY', p..':enqueue:'..queue, 'entered', 1)
  if state == 'scheduled' then redis.call('ZADD', p..':sched', sched, id) end
  -- control plane inspection indexes: per-state zset, live-jobs-by-fingerprint set, and the
  -- backlog metrics per-minute arrival counter (TTL ~25h — history is a window, not an archive).
  redis.call('ZADD', p..':idx:'..queue..':'..state, sched, id)
  if state == 'available' then redis.call('ZADD', p..':avail:'..queue..':'..part, sched, id) end
  local fp = ARGV[b + 7]
  if fp ~= '' then redis.call('SADD', p..':fpi:'..fp, id) end
  for _, tag in ipairs(cjson.decode(tagsjson)) do redis.call('SADD', p..':tag:'..tag, id) end
  local hb = p..':hist:'..queue..':'..tostring(now - now % 60000)
  redis.call('HINCRBY', hb, 'arrived', 1)
  redis.call('PEXPIRE', hb, 90000000)
  local hbp = p..':histp:'..queue..':'..part..':'..tostring(now - now % 60000)
  redis.call('HINCRBY', hbp, 'arrived', 1)
  redis.call('PEXPIRE', hbp, 90000000)
  local uk, uw = ARGV[b + 14], tonumber(ARGV[b + 15])
  if uk ~= '' then
    if uw > 0 then redis.call('SET', p..':uniquet:'..uk, id, 'PX', uw)
    else redis.call('SET', p..':unique:'..uk, id) end
  end
  -- backend wakeup contract push wakeup, once per distinct queue — the Redis twin of enqueue's pg_notify.
  -- A dropped message costs latency, never correctness (the poll fallback stands).
  if not woken[queue] then
    woken[queue] = true
    redis.call('PUBLISH', p..':wake', queue)
  end
 end
end
return {'OK', tostring(written)}

File: go/driver/headgatemysql/enqueue.go
package headgatemysql

import (
	"context"
	"database/sql"
	"errors"
	"slices"
	"strings"

	headgate "github.com/mujhtech/headgate/go"
)

// Enqueue inserts a batch of jobs atomically.
func (s *MysqlStore) Enqueue(ctx context.Context, batch []headgate.Envelope) error {
	if len(batch) == 0 {
		return nil
	}
	// Validate before acquiring a connection. An invalid envelope remains a caller error
	// even when MySQL is unreachable; transport state must not change the taxonomy.
	if err := headgate.ValidateEnqueue(batch); err != nil {
		return err
	}
	scoped := append([]headgate.Envelope(nil), batch...)
	for i := range scoped {
		scoped[i].UniqueKey = headgate.EffectiveUniqueKey(scoped[i])
	}
	batch = scoped
	// the rows and their active-partition entries must land together: a crash
	// between them would leave an available job whose partition is not listed, which is
	// starvation. (EnqueueTx already supplies the caller's own transaction.)
	tx, err := s.db.BeginTx(ctx, &sql.TxOptions{Isolation: sql.LevelReadCommitted})
	if err != nil {
		return headgate.WrapUnavailable(err)
	}
	defer tx.Rollback() //nolint:errcheck // no-op after a successful Commit
	if err := s.enqueueOn(ctx, tx, batch); err != nil {
		var duplicate *headgate.DuplicateError
		if errors.As(err, &duplicate) && duplicate.Replaced {
			if commitErr := tx.Commit(); commitErr != nil {
				return headgate.WrapUnavailable(commitErr)
			}
			return err
		}
		return headgate.WrapUnavailable(err)
	}
	return headgate.WrapUnavailable(tx.Commit())
}

// execer is what enqueue needs from either *sql.DB or *sql.Tx.
type execer interface {
	ExecContext(context.Context, string, ...any) (sql.Result, error)
	QueryRowContext(context.Context, string, ...any) *sql.Row
	QueryContext(context.Context, string, ...any) (*sql.Rows, error)
}

func (s *MysqlStore) enqueueOn(ctx context.Context, c execer, batch []headgate.Envelope) error {
	if len(batch) == 0 {
		return nil
	}
	// typed dispatch / boundary validation / idempotent enqueue identity one shared boundary check for every backend.
	if err := headgate.ValidateEnqueue(batch); err != nil {
		return err
	}
	// idempotent enqueue identity the strict caller-supplied id contract, classified BEFORE anything is written
	// so the batch stays all-or-nothing (and this whole function already runs inside a
	// transaction, plain or caller-supplied). Matching content drops out — idempotent
	// success, what makes the API's Idempotency-Key replay safe; different content
	// rejects the whole batch naming the offender. A terminal row still counts as
	// existing; reuse follows retention eviction.
	{
		ids := make([]any, len(batch))
		for i, e := range batch {
			ids[i] = e.ID
		}
		present := map[string][3]string{}
		rows, err := c.QueryContext(ctx,
			`SELECT ulid, kind, fingerprint, queue FROM headgate_job WHERE ulid IN (`+
				placeholders(len(batch))+`)`, ids...)
		if err != nil {
			return err
		}
		for rows.Next() {
			var id, kind, fp, queue string
			if err := rows.Scan(&id, &kind, &fp, &queue); err != nil {
				_ = rows.Close()
				return err
			}
			present[id] = [3]string{kind, fp, queue}
		}
		_ = rows.Close()
		if err := rows.Err(); err != nil {
			return err
		}
		kept := make([]headgate.Envelope, 0, len(batch))
		for _, e := range batch {
			ex, exists := present[e.ID]
			switch {
			case !exists:
				kept = append(kept, e)
			case headgate.SameJobContent(e, ex[0], ex[1], ex[2]):
				// idempotent replay — do not re-write it, do not duplicate it
			default:
				return &headgate.IDConflictError{JobID: e.ID}
			}
		}
		if len(kept) == 0 {
			return nil // every row already exists, unchanged — nothing to write
		}
		batch = kept
	}
	// crash quarantine quarantined fingerprints are rejected before anything is written.
	var fps []any
	for _, e := range batch {
		if e.Fingerprint != "" {
			fps = append(fps, e.Fingerprint)
		}
	}
	if len(fps) > 0 {
		var fp string
		err := c.QueryRowContext(ctx,
			"SELECT fingerprint FROM headgate_quarantine WHERE fingerprint IN ("+
				placeholders(len(fps))+") LIMIT 1", fps...).Scan(&fp)
		if err == nil {
			return &headgate.QuarantinedError{Fingerprint: fp}
		}
		if !errors.Is(err, sql.ErrNoRows) {
			return err
		}
	}
	// Exact producer depth lives in two PK counter rows. Lock policy rows in sorted
	// order so concurrent multi-queue batches cannot overrun or deadlock. Idempotent
	// replays were removed above and therefore consume no slot.
	demand := map[string]int64{}
	for _, e := range batch {
		demand[headgate.EnqueueQueue(e)]++
	}
	demandQueues := make([]string, 0, len(demand))
	for queue := range demand {
		demandQueues = append(demandQueues, queue)
	}
	slices.Sort(demandQueues)
	policyArgs := make([]any, len(demandQueues))
	for i, queue := range demandQueues {
		policyArgs[i] = queue
	}
	if _, err := c.ExecContext(ctx,
		`INSERT INTO headgate_enqueue_policy (queue) VALUES `+
			strings.TrimSuffix(strings.Repeat("(?), ", len(demandQueues)), ", ")+`
		 AS new ON DUPLICATE KEY UPDATE queue = new.queue`, policyArgs...); err != nil {
		return err
	}
	// Materialize both counter rows before the locking read. A LEFT JOIN FOR UPDATE
	// against absent rows takes next-key gap locks; producers for different queues can
	// then each hold the same gap and deadlock when the INSERT trigger creates `entered`.
	// Sorted PK inserts make the query below take record locks instead.
	counterArgs := make([]any, 0, len(demandQueues)*2)
	for _, queue := range demandQueues {
		counterArgs = append(counterArgs, queue, queue)
	}
	if _, err := c.ExecContext(ctx,
		`INSERT IGNORE INTO headgate_enqueue_counter (queue, counter_kind, n) VALUES `+
			strings.TrimSuffix(strings.Repeat("(?, 'entered', 0), (?, 'exited', 0), ", len(demandQueues)), ", "),
		counterArgs...); err != nil {
		return err
	}
	// The sorted no-op upsert takes policy X locks immediately. The counter SELECT is
	// still a locking/current read over real records: a caller's REPEATABLE READ
	// transaction may have established its snapshot in the earlier ID pre-check.
	rows, err := c.QueryContext(ctx, enqueueBackpressureDepthSQL(len(demandQueues)), policyArgs...)
	if err != nil {
		return err
	}
	for rows.Next() {
		var queue string
		var limit sql.NullInt64
		var entered, exited int64
		if err := rows.Scan(&queue, &limit, &entered, &exited); err != nil {
			_ = rows.Close()
			return err
		}
		if limit.Valid {
			current := max(int64(0), entered-exited)
			incoming := demand[queue]
			if current+incoming > limit.Int64 {
				_ = rows.Close()
				return &headgate.BackpressureError{Queue: queue, Limit: uint64(limit.Int64), Current: uint64(current), Incoming: uint64(incoming)}
			}
		}
	}
	_ = rows.Close()
	if err := rows.Err(); err != nil {
		return err
	}
	var now int64
	if err := c.QueryRowContext(ctx, "SELECT "+nowMS).Scan(&now); err != nil {
		return err
	}
	rowSQL := `(?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ` + nowMS + `,
	            ?, ?, ?, ?, ?, JSON_ARRAY(), ?, ?, ?, ?, CAST(? AS JSON), ?, ?, ?)`
	values := strings.TrimSuffix(strings.Repeat(rowSQL+", ", len(batch)), ", ")
	stmt := `INSERT INTO headgate_job
	           (ulid, kind, schema_version, payload, queue, partition_key, rate_class,
	            weight, fingerprint, priority, max_attempts, enqueued_at_ms,
	            scheduled_at_ms, timeout_ms, deadline_ms, retention_ms,
	            state, errors, unique_key, unique_states, unique_window_ms,
	            unique_expires_at_ms, headers, periodic_schedule_id, periodic_tick_ms,
	            sticky_worker) VALUES ` + values
	var args []any
	var candidates []any
	for _, e := range batch {
		sv := headgate.EffectiveSchemaVersion(e.SchemaVersion)
		queue := headgate.EnqueueQueue(e)
		ma := headgate.EffectiveMaxAttempts(e.MaxAttempts)
		var expires any
		if e.UniqueKey != nil && e.UniqueWindowMs > 0 {
			expires = now + e.UniqueWindowMs
		}
		scheduledAt := e.ScheduledAtMs
		if e.UniqueDebounceMs > 0 {
			scheduledAt = now + e.UniqueDebounceMs
		} else if scheduledAt == 0 {
			scheduledAt = now
		}
		state := "available"
		if e.Pending {
			state = "pending"
		} else if scheduledAt > now {
			state = "scheduled"
		}
		args = append(args,
			e.ID, e.Kind, sv, e.Payload, queue, e.PartitionKey, e.RateClass,
			headgate.EffectiveWeight(e.Weight), e.Fingerprint, e.Priority, ma,
			scheduledAt, e.TimeoutMs, e.DeadlineMs, e.RetentionMs, state,
			e.UniqueKey, e.UniqueStates, e.UniqueWindowMs, expires,
			// Opaque headers are encoded without being interpreted.
			headersJSON(e.Headers), e.PeriodicScheduleID, e.PeriodicTickMs, e.StickyWorker)
		if e.UniqueKey != nil {
			candidates = append(candidates, e.UniqueKey)
		}
	}

	// Lock maintained route rows before inserting jobs. Pruning and reconciliation take
	// route -> job; the old job -> route producer order formed an InnoDB deadlock. A
	// harmless stale route can be pruned, while an available job without one can starve.
	routes := make([]string, 0, len(batch))
	for _, e := range batch {
		queue := headgate.EnqueueQueue(e)
		routes = append(routes, queue+"\x00"+e.PartitionKey)
	}
	slices.Sort(routes)
	routes = slices.Compact(routes)
	routeArgs := make([]any, 0, len(routes)*2)
	for _, route := range routes {
		queue, partition, _ := strings.Cut(route, "\x00")
		routeArgs = append(routeArgs, queue, partition)
	}
	routeValues := strings.TrimSuffix(strings.Repeat("(?, ?), ", len(routes)), ", ")
	if _, err := c.ExecContext(ctx,
		`INSERT INTO headgate_active_partition (queue, partition_key)
		 VALUES `+routeValues+`
		 ON DUPLICATE KEY UPDATE queue = VALUES(queue)`, routeArgs...); err != nil {
		return err
	}
	inflightValues := strings.TrimSuffix(strings.Repeat("(?, ?, 0), ", len(routes)), ", ")
	if _, err := c.ExecContext(ctx,
		`INSERT INTO headgate_inflight (queue, partition_key, n)
		 VALUES `+inflightValues+`
		 AS new ON DUPLICATE KEY UPDATE n = headgate_inflight.n`, routeArgs...); err != nil {
		return err
	}

	for attempt := 0; attempt < 2; attempt++ {
		_, err := c.ExecContext(ctx, stmt, args...)
		if err == nil {
			for _, e := range batch {
				for _, tag := range headgate.CanonicalTags(e.Tags) {
					if _, err := c.ExecContext(ctx, `INSERT INTO headgate_job_tag(job_id,tag) SELECT id,? FROM headgate_job WHERE ulid=?`, tag, e.ID); err != nil {
						return err
					}
				}
			}
			// backlog metrics arrived counters, one upsert per distinct queue.
			perQueue := map[string]int64{}
			for _, e := range batch {
				queue := e.Queue
				if queue == "" {
					queue = "default"
				}
				perQueue[queue]++
			}
			for q, n := range perQueue {
				if _, err := c.ExecContext(ctx,
					`INSERT INTO headgate_queue_counter (queue, bucket_ms, arrived)
					 VALUES (?, (`+nowMS+` DIV 60000) * 60000, ?) AS new
					 ON DUPLICATE KEY UPDATE arrived = headgate_queue_counter.arrived + new.arrived`,
					q, n); err != nil {
					return err
				}
			}
			perPartition := map[string]int64{}
			for _, e := range batch {
				queue := e.Queue
				if queue == "" {
					queue = "default"
				}
				perPartition[queue+"\x00"+e.PartitionKey]++
			}
			for route, n := range perPartition {
				queue, partition, _ := strings.Cut(route, "\x00")
				if _, err := c.ExecContext(ctx,
					`INSERT INTO headgate_partition_counter
					   (queue, partition_key, bucket_ms, arrived)
					 VALUES (?, ?, (`+nowMS+` DIV 60000) * 60000, ?) AS new
					 ON DUPLICATE KEY UPDATE arrived = headgate_partition_counter.arrived + new.arrived`,
					queue, partition, n); err != nil {
					return err
				}
			}
			return nil
		}
		if !isDupKey(err) {
			return err
		}
		// Throttle keys release LAZILY: the conflicting enqueue clears any holder
		// whose window has passed, then retries once.
		if attempt == 0 && len(candidates) > 0 {
			res, rerr := c.ExecContext(ctx,
				lazyUniqueReleaseSQL(len(candidates)), candidates...)
			if rerr == nil {
				if released, _ := res.RowsAffected(); released > 0 {
					continue
				}
			}
		}
		// job uniqueness one semantic: the duplicate is a normal result carrying the winner's
		// id — never a silent skip, never a bare constraint error.
		if len(candidates) > 0 {
			var existing string
			holderArgs := append(append([]any{}, candidates...), candidates...)
			err := c.QueryRowContext(ctx,
				uniqueHolderSQL(len(candidates)), holderArgs...).Scan(&existing)
			if err == nil {
				incoming := batch[0]
				replaced := false
				if incoming.UniqueDebounceMs > 0 {
					schemaVersion := incoming.SchemaVersion
					if schemaVersion == 0 {
						schemaVersion = 1
					}
					res, updateErr := c.ExecContext(ctx, `UPDATE headgate_job SET schema_version=?,payload=?,fingerprint=?,state='scheduled',scheduled_at_ms=`+nowMS+`+? WHERE ulid=? AND state IN ('pending','scheduled','available','retryable')`, schemaVersion, incoming.Payload, incoming.Fingerprint, incoming.UniqueDebounceMs, existing)
					if updateErr != nil {
						return updateErr
					}
					rowsAffected, _ := res.RowsAffected()
					replaced = rowsAffected > 0
					if replaced {
						if _, err := c.ExecContext(ctx, `DELETE FROM headgate_job_tag WHERE job_id=(SELECT id FROM headgate_job WHERE ulid=?)`, existing); err != nil {
							return err
						}
						for _, tag := range headgate.CanonicalTags(incoming.Tags) {
							if _, err := c.ExecContext(ctx, `INSERT INTO headgate_job_tag(job_id,tag) SELECT id,? FROM headgate_job WHERE ulid=?`, tag, existing); err != nil {
								return err
							}
						}
					}
				} else if incoming.UniqueReplace != 0 {
					schemaVersion := incoming.SchemaVersion
					if schemaVersion == 0 {
						schemaVersion = 1
					}
					maxAttempts := incoming.MaxAttempts
					if maxAttempts == 0 {
						maxAttempts = 25
					}
					res, updateErr := c.ExecContext(ctx, `UPDATE headgate_job SET
						schema_version = IF((? & ?) <> 0, ?, schema_version),
						payload = IF((? & ?) <> 0, ?, payload), fingerprint = IF((? & ?) <> 0, ?, fingerprint),
						scheduled_at_ms = IF((? & ?) <> 0 AND state = 'scheduled', IF(? = 0, `+nowMS+`, ?), scheduled_at_ms),
						priority = IF((? & ?) <> 0, ?, priority), max_attempts = IF((? & ?) <> 0, ?, max_attempts)
					  WHERE ulid = ? AND state IN ('scheduled','available','retryable')
					    AND ((? & ?) <> 0 OR (? & ?) <> 0 OR (? & ?) <> 0 OR ((? & ?) <> 0 AND state = 'scheduled'))`,
						incoming.UniqueReplace, headgate.UniqueReplacePayload, schemaVersion,
						incoming.UniqueReplace, headgate.UniqueReplacePayload, incoming.Payload,
						incoming.UniqueReplace, headgate.UniqueReplacePayload, incoming.Fingerprint,
						incoming.UniqueReplace, headgate.UniqueReplaceScheduledAt, incoming.ScheduledAtMs, incoming.ScheduledAtMs,
						incoming.UniqueReplace, headgate.UniqueReplacePriority, incoming.Priority,
						incoming.UniqueReplace, headgate.UniqueReplaceMaxAttempts, maxAttempts, existing,
						incoming.UniqueReplace, headgate.UniqueReplacePayload, incoming.UniqueReplace, headgate.UniqueReplacePriority,
						incoming.UniqueReplace, headgate.UniqueReplaceMaxAttempts, incoming.UniqueReplace, headgate.UniqueReplaceScheduledAt)
					if updateErr != nil {
						return updateErr
					}
					rowsAffected, _ := res.RowsAffected()
					replaced = rowsAffected > 0
				}
				return &headgate.DuplicateError{ExistingID: existing, Replaced: replaced}
			}
			if !errors.Is(err, sql.ErrNoRows) {
				return err
			}
		}
		// Not a uniqueness index — the ulid key collided. The pre-check above classified
		// every id this call knew about, so reaching here means a CONCURRENT producer
		// inserted the row between the read and the write; idempotent enqueue identity's answer is the same
		// typed conflict, naming the id.
		ids := make([]any, len(batch))
		for i, e := range batch {
			ids[i] = e.ID
		}
		var raced string
		if rerr := c.QueryRowContext(ctx,
			`SELECT ulid FROM headgate_job WHERE ulid IN (`+placeholders(len(batch))+
				`) LIMIT 1`, ids...).Scan(&raced); rerr != nil && !errors.Is(rerr, sql.ErrNoRows) {
			return rerr
		}
		return &headgate.IDConflictError{JobID: raced}
	}
	panic("enqueue retries at most once")
}

// ReclaimExpired makes jobs with expired leases eligible for admission again.
